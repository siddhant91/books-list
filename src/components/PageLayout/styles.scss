@import '../../styles/main.scss';

.bokl-page-layout {
	display: flex;
	box-sizing: border-box;
	&__header {
		height: $bokl-spacing-8;
		overflow: hidden;
		position: fixed;
		top: 0;
		width: 100%;
		z-index: 10;
		background-color: $bokl-color-white;
		box-shadow: 0 5px 20px -10px $bokl-color-black;
		&--logo {
			width: 100px;
			height: 34px;
		}
	}
	&__main {
		display: flex;
		width: 100%;
		flex-direction: column;
		overflow: hidden;
	}
	&__content {
		margin-top: $bokl-spacing-8;
		padding: $bokl-spacing-3;
		background-color: $bokl-color-white-1;
	}
}
