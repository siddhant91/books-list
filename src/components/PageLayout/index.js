import clsx from 'clsx';
import { string, node, arrayOf, shape } from 'prop-types';
import { useState } from 'react';
// Logo Assets
import LogoImage from '../../assets/images/Logo.png';

// Compoents
import Button from '../Button';

// Styles
import './styles.scss';

const PageLayout = ({ additonalClasses, menuItems, children, ...rest }) => {
	const [isMenuOpen, toggleMenuState] = useState(false);
	const toggleMenu = () => {
		toggleMenuState(!isMenuOpen);
	};
	const getMenuItems = () => {
		const items = menuItems.map((item) => {
			const { title, icon, isAccordion } = item;

			return (
				<Button
					aria-label={title}
					appearance="just-icon"
					size="icon"
					additionalClasses="text-left p-0 d-flex align-items-center mt-3"
					key={title}
					data-testid="menu-item-btn"
					handleClick={(e) => {
						console.log(e);
					}}
				>
					<i className={`fa ${icon}`} aria-hidden="true" />

					<span
						className={clsx('bokl-page-layout__menu--item-title ', {
							'd-flex justify-content-between align-items-center flex-grow-1': isMenuOpen,
						})}
					>
						{title}
						{isAccordion && <i className="fa fa-angle-down" aria-hidden="true" />}
					</span>
				</Button>
			);
		});
		return items;
	};
	const renderMenu = () => {
		return (
			<div
				className={clsx('bokl-page-layout__menu', {
					open: isMenuOpen,
				})}
				data-testid="menu-container"
			>
				<Button
					aria-label="open Menu"
					appearance="just-icon"
					size="icon"
					additionalClasses="text-left p-0 d-flex align-items-center"
					handleClick={toggleMenu}
					data-testid="toggle-menu"
				>
					<i className="fa fa-bars" aria-hidden="true" />
					<span className="bokl-page-layout__menu--item-title">Menu</span>
				</Button>
				{menuItems.length > 0 && (
					<div className="d-flex flex-column mt-3" data-testid="menu-items">
						{getMenuItems()}
					</div>
				)}
			</div>
		);
	};
	return (
		<div className={clsx('bokl-page-layout', additonalClasses)} {...rest}>
			{renderMenu()}
			<div
				className={clsx('bokl-page-layout__main', {
					'shift-left': isMenuOpen,
				})}
			>
				<header className="bokl-page-layout__header d-flex align-items-center">
					<img
						className="bokl-page-layout__header--logo"
						src={LogoImage}
						alt="Logo"
						data-testid="logo-img"
					/>
				</header>
				<main className="bokl-page-layout__content">{children}</main>
			</div>
		</div>
	);
};

PageLayout.propTypes = {
	additonalClasses: string,
	menuItems: arrayOf(shape({})).isRequired,
	children: node,
};

PageLayout.defaultProps = {
	additonalClasses: '',
	children: null,
};

export default PageLayout;
