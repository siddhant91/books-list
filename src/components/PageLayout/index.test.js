import { render, cleanup, waitFor, fireEvent } from '@testing-library/react';

import PageLayout from './index';

afterEach(cleanup);

const mockMenuItems = [
	{
		title: 'Content Management',
		icon: 'fa-window-maximize',
		isAccordion: true,
	},
	{
		title: 'Courses',
		icon: 'fa-file',
		isAccordion: false,
	},
];

const renderComponent = (menuItems = []) => {
	return (
		<PageLayout menuItems={menuItems}>
			<p>this is child node text</p>
		</PageLayout>
	);
};
it('Matches snapshot', () => {
	const { asFragment } = render(renderComponent());
	expect(asFragment()).toMatchSnapshot();
});

it('Renders Logo Img', () => {
	const { getByTestId } = render(renderComponent());
	expect(getByTestId('logo-img')).toBeInTheDocument();
});

it('Renders Chidren', async () => {
	const { findByText } = render(renderComponent());
	expect(await findByText('this is child node text')).toBeInTheDocument();
});

it('Renders Menu Items', async () => {
	const { findByTestId } = render(renderComponent(mockMenuItems));

	const menuElement = await findByTestId('menu-items');
	await waitFor(() => {
		expect(menuElement).toBeInTheDocument();
	});
});

it('toggles the menu State', async () => {
	const { getByTestId } = render(renderComponent());
	const menuToggler = getByTestId('toggle-menu');
	const menuContainer = getByTestId('menu-container');
	fireEvent.click(menuToggler);
	await waitFor(() => {
		expect(menuContainer.classList.contains('open')).toBe(true);
	});
});
