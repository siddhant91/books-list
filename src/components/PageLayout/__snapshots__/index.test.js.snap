// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bokl-page-layout"
  >
    <div
      class="bokl-page-layout__menu"
      data-testid="menu-container"
    >
      <button
        aria-label="open Menu"
        class="text-left p-0 d-flex align-items-center just-icon icon bokl-button-component"
        data-testid="toggle-menu"
        type="button"
      >
        <i
          aria-hidden="true"
          class="fa fa-bars"
        />
        <span
          class="bokl-page-layout__menu--item-title"
        >
          Menu
        </span>
      </button>
    </div>
    <div
      class="bokl-page-layout__main"
    >
      <header
        class="bokl-page-layout__header d-flex align-items-center"
      >
        <img
          alt="Logo"
          class="bokl-page-layout__header--logo"
          data-testid="logo-img"
          src="Logo.png"
        />
      </header>
      <main
        class="bokl-page-layout__content"
      >
        <p>
          this is child node text
        </p>
      </main>
    </div>
  </div>
</DocumentFragment>
`;
